include/s3select_json_parser.h

```diff
@@ -0,0 +1,419 @@
+    bool prefix_match{};
+    Valuesax value;
+    ChunksStreamer stream_buffer;
+    bool init_buffer_stream;
+    rapidjson::Reader reader;
+    std::vector<en_json_elm_state_t> json_element_state;
+    std::string m_result;//debug purpose
+    std::vector<std::string> key_path;
+
+    JsonParserHandler() : init_buffer_stream(false)
+    {} 
+
+    std::string get_key_path()
+    {//for debug
+	  std::string res;
+	  for(const auto & i: key_path)
+	  {
+	    res.append(i);
+	    res.append(std::string("/"));
+	  }
+	  return res;
+    }
+
+    void dec_key_path()
+    {
+      if(json_element_state.back() != ARRAY_STATE)
+	    {
+	      if(key_path.size() != 0) {
+	      key_path.pop_back();
+        }
+	    }
+      if (key_path.size() < from_clause.size()) {
+        prefix_match = false;
+      } 
+      if (prefix_match) {
+          if (state == row_state::ARRAY || state == row_state::START_ROW) {
+            state = row_state::START_ROW;
+            ++row_count;
+          }
+      }
+    }
+
+    void push_new_key_value(Valuesax& v)
+    {  
+      if (prefix_match) {
+        for (size_t i = 1; i < query_matrix.size(); i++) { // 0th index -> from-clause. Ignore that.
+          if(std::equal(key_path.begin() + from_clause.size(), key_path.end(), query_matrix[i].begin())) {
+            fp(make_pair(get_key_path(),v));
+          }
+        }
+      }
+      dec_key_path();
+    }
+
+    bool Null() {
+      // at this point should verify against from-clause/where-clause/project. if match then push to scratch-area
+      push_new_key_value(value.Parse(nullptr));
+      return true; }
+
+    bool Bool(bool b) {
+      push_new_key_value(value.Parse(b));
+      return true; }
+
+    bool Int(int i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint(unsigned u) {
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Int64(int64_t i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint64(uint64_t u) { 
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Double(double d) { 
+      push_new_key_value(value.Parse(d));
+      return true; }
+
+    bool String(const char* str, rapidjson::SizeType length, bool copy) {
+      //TODO use copy
+      push_new_key_value(value.Parse(str));
+      return true;
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool copy) {
+      key_path.push_back(std::string(str));
+      
+      if (key_path == from_clause) {
+        prefix_match = true;
+      }
+      return true;
+    }
+
+    bool StartObject() {      
+      if (key_path.size()) {
+        if (key_path[key_path.size() -1] == from_clause[from_clause.size() - 1]) {
+          if (state != row_state::ARRAY && state != row_state::ARRAY_END_ROW) {
+          state = row_state::OBJECT_START_ROW;
+          ++row_count;
+        } else {
+          state = row_state::ARRAY_START_ROW;
```

Comment:
why setting `ARRAY_START_ROW` upon `StartObject` is been called ?

Reply:
Since there are 2 use cases for array, array of objects and normal array and since their behaviour  is quite different, 2 states are used to indicate it